### ISSUE TITLE
10-Fold Cross Validation Code and alternative metrics?

### ISSUE THREAD

**Issue by ieee8023** [dmlc/mxnet#1252]
Not sure if this is the correct place to post this. 

Is there some existing code that will do 10-Fold Cross Validation with the RecordIO data? My current thought is to generate many RecordIO files and use them for training and testing.

Examples show the accuracy on a validation set but I'm not sure how to open this up and calculate something like the F1-Score or Precision and Recall. Any pointers?

**Comment by piiswrong** [comment]
RecordIO supports chunking. Split the data into 10 chunks and use 9 for training and 1 for testing ever time.

**Comment by ieee8023** [comment]
@piiswrong Thanks! You mean using num_parts in the DataIter? I'm not sure how to grab parts 1-9. 

```
train = mx.io.ImageRecordIter(
    path_imgrec = "data.rec",
    ...
    num_parts   = 10,
    part_index = 0
)
```

**Comment by piiswrong** [comment]
You probably need to write a python DataIter to wrap multiple ImageRecordIters on different parts.

**Comment by ieee8023** [comment]
I'm closing this issue. I couldn't figure out the data iterator but I wrote an EvalMetric to calculate the F1-Score in case people are looking for it in the future. It is hard coded to a specific class label though.

```
# coding: utf-8
from __future__ import absolute_import
from mxnet import metric
"""Online evaluation metric module."""

# from .base import string_types
import numpy

class F1(metric.EvalMetric):
    """Calculate F1-Score in terms of label 1"""
    def __init__(self):
        self.focusLabel = 1;
        super(F1, self).__init__('f1')

    def reset(self):
        self.tp = 0.0
        self.tn = 0.0
        self.fp = 0.0
        self.fn = 0.0
        super(F1, self)
        
    def get(self):
        try:
            precision = self.tp / (self.tp + self.fp)
            recall = self.tp / (self.tp + self.fn)
            return (self.name, 2 * (precision * recall) / (precision + recall))
        except:
            return (self.name, 0.0)

    def update(self, labels, preds):
        assert len(labels) == len(preds)
        for i in range(len(labels)):
            pred = preds[i].asnumpy()
            label = labels[i].asnumpy().astype('int32')
            pred_label = numpy.argmax(pred, axis=1)
            if label.shape[0] < pred_label.shape[0]:
                raise Exception("Predict label is more than data label? ")
            ### calc f1 here???!!!
            correct = pred_label == label[:pred_label.shape[0]]
            incorrect = pred_label != label[:pred_label.shape[0]]
            pos = pred_label == self.focusLabel
            neg = pred_label != self.focusLabel
            self.tp += numpy.sum(correct & pos)
            self.tn += numpy.sum(correct & neg)
            self.fp += numpy.sum(incorrect & pos)
            self.fn += numpy.sum(incorrect & neg)
            #print("tp = " + `self.tp`)
            #print("tn = " + `self.tn`)
            #print("fp = " + `self.fp`)
            #print("fn = " + `self.fn`)
```

It gets used like this:

```
    model.fit(
        X                  = train,
        eval_data          = val,
        kvstore            = kv,
        batch_end_callback = mx.callback.Speedometer(args.batch_size, 50),
        epoch_end_callback = checkpoint,
        eval_metric = metric_f1.F1())
```
